man8/thin_dump.8
.TH THIN_DUMP 8 "Thin Provisioning Tools" "Red Hat, Inc." \" -*- nroff -*-
.SH NAME
thin_dump \- dump thin provisioning metadata from device or file to standard output

.SH SYNOPSIS
.B thin_dump
.RB [options]
.I {device|file}

.SH DESCRIPTION
.B thin_dump
dumps binary thin provisioning metadata (optionally from alternate block;
see option \-\-metadata-snap) created by the device-mapper
thin provisioning target on a
.I device
or
.I file
to standard output for
analysis or postprocessing in either XML or human readable format.
XML formated metadata can be fed into thin_restore (see
.BR thin_restore(8) )
in order to put it back onto a metadata
.I device
(to process by the device-mapper target) or
.I file.

This tool cannot be run on live metadata unless the \fB\-\-metadata\-snap\fP option is used.

.IP "\fB\-f, \-\-format\fP \fI{xml|human_readable|custom}\fP".

Print output in XML or human readable format.  Custom formats are
supported via shared library plugins.  They should be specified as in
this example:
.sp
.B thin_dump --format custom=mylib.so /dev/sda

.IP "\fB\-r, \-\-repair\fP".
Repair the metadata whilst dumping it.

.IP "\fB\-m, \-\-metadata\-snap\fP [block#]".
Dump metadata snapshot created by device-mapper thin provisioning target.
If block is not provided, access the default metadata snapshot created by
the thin provisioning device-mapper target, else try the one at block#.
See the thin provisioning target documentation on how to create or release
a metadata snapshot and retrieve the block number from the kernel.

.IP "\fB\-\-dev\-id\fP <dev-id>".
Dump the specified device.  This option may be specified multiple
times to select more than one thin device.

.IP "\fB\-\-skip\-mappings".
Do not dump the mappings.

.IP "\fB-o \fI{xml file}\fP"
Specify a file for the output rather than writing to stdout.

.IP "\fB\-h, \-\-help\fP".
Print help and exit.

.IP "\fB\-V, \-\-version\fP".
Output version information and exit.

.SH EXAMPLES
Dumps the thin provisioning metadata on logical volume /dev/vg/metadata
to standard output in human readable format:
.sp
.B thin_dump -f human_readable /dev/vg/metadata

Dumps the thin provisioning metadata on logical volume /dev/vg/metadata
to standard output in XML format:
.sp
.B thin_dump /dev/vg/metadata

Dumps the thin provisioning metadata snapshot on logical volume /dev/vg/metadata
to standard output in human readable format (not processable by
.B thin_restore(8)
):
.sp
.B thin_dump --format human_readable --metadata-snap /dev/vg/metadata

.SH DIAGNOSTICS
.B thin_dump
returns an exit code of 0 for success or 1 for error.

.SH SEE ALSO
.B thin_check(8)
.B thin_repair(8)
.B thin_restore(8)
.B thin_rmap(8)
.B thin_metadata_size(8)
.SH AUTHOR
Joe Thornber <ejt@redhat.com>
.br
Heinz Mauelshagen <HeinzM@RedHat.com>


CHANGES
v0.6.2
======

- Fix bug in thin_delta
- Fix recent regression in thin_repair.
- Force g++-98 dialect
- Fix bug in thin_trim

v0.6.1
======

Fix bug in thin_dump when using metadata snaps.

v0.6.0
======

thin_ls

v0.5.6
======

era_invalidate may be run on live metadata if the --metadata-snap option is given.

v0.5.5
======

You may now give the --metadata_snap option to thin_delta without specifying where the snap is.

Update man pages to make it clearer that most tools shouldn't be run on live metadata.

Fix some bugs in the metadata reference counting for thin_check.

v0.5.1
======

Fix bug where the tools would crash if given a very large metadata
device to restore to.

v0.5
====

- thin_delta, thin_trim
- --clear-needs-check flag for cache_check
- space map checking for thin check

v0.4
====

- All tools switch to using libaio.  This gives a large performance
  boost, especially to the write focused tools like thin_restore.

- Added a progress monitor to thin_restore, cache_restore and era_restore

- Added a --quiet/-q option to *_restore to turn off the progress bar

- Removed variable hint size support from cache tools.  The kernel
  still only supports a fixed 32bit width.  This will have a side
  effect of reducing the executable sizes due to less template
  instatiation.

- Tools rolled into a single executable to save space.

- Fixed some bugs when walking bitsets (possibly effecting cache_dump
  and cache_check).

thin-provisioning/thin_dump.cc
// Copyright (C) 2011 Red Hat, Inc. All rights reserved.
//
// This file is part of the thin-provisioning-tools source.
//
// thin-provisioning-tools is free software: you can redistribute it
// and/or modify it under the terms of the GNU General Public License
// as published by the Free Software Foundation, either version 3 of
// the License, or (at your option) any later version.
//
// thin-provisioning-tools is distributed in the hope that it will be
// useful, but WITHOUT ANY WARRANTY; without even the implied warranty
// of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License along
// with thin-provisioning-tools.  If not, see
// <http://www.gnu.org/licenses/>.

#include <fstream>
#include <iostream>
#include <getopt.h>
#include <libgen.h>

#include "persistent-data/file_utils.h"
#include "thin-provisioning/commands.h"
#include "thin-provisioning/human_readable_format.h"
#include "thin-provisioning/metadata.h"
#include "thin-provisioning/metadata_dumper.h"
#include "thin-provisioning/shared_library_emitter.h"
#include "thin-provisioning/xml_format.h"
#include "version.h"

using namespace boost;
using namespace persistent_data;
using namespace std;
using namespace thin_provisioning;

namespace {
	// FIXME: put the path into the flags
	struct flags {
		flags()
			: format("xml"),
			  use_metadata_snap(false) {
		}

		dump_options opts;

		string format;
		bool use_metadata_snap;
		optional<block_address> snap_location;
	};

	metadata::ptr open_metadata(string const &path, struct flags &flags) {
		block_manager<>::ptr bm = open_bm(path, block_manager<>::READ_ONLY, !flags.use_metadata_snap);
		metadata::ptr md(flags.use_metadata_snap ? new metadata(bm, flags.snap_location) : new metadata(bm));

		return md;
	}

	bool begins_with(string const &str, string const &prefix) {
		return str.substr(0, prefix.length()) == prefix;
	}

	emitter::ptr create_emitter(string const &format, ostream &out) {
		emitter::ptr e;

		if (format == "xml")
			e = create_xml_emitter(out);

		else if (format == "human_readable")
			e = create_human_readable_emitter(out);

		else if (begins_with(format, "custom="))
			e = create_custom_emitter(format.substr(7), out);

		else {
			ostringstream msg;
			msg << "unknown format '" << format << "'";
			throw runtime_error(msg.str());
		}

		return e;
	}

	int dump_(string const &path, ostream &out, struct flags &flags) {
		try {
			metadata::ptr md = open_metadata(path, flags);
			emitter::ptr e = create_emitter(flags.format, out);
			metadata_dump(md, e, flags.opts);

		} catch (std::exception &e) {
			cerr << e.what() << endl;
			return 1;
		}

		return 0;
	}

	int dump(string const &path, char const *output, struct flags &flags) {
		if (output) {
			ofstream out(output);
			return dump_(path, out, flags);
		} else
			return dump_(path, cout, flags);
	}
}

//----------------------------------------------------------------

thin_dump_cmd::thin_dump_cmd()
	: command("thin_dump")
{
}

void
thin_dump_cmd::usage(std::ostream &out) const
{
	out << "Usage: " << get_name() << " [options] {device|file}\n"
	    << "Options:\n"
	    << "  {-h|--help}\n"
	    << "  {-f|--format} {xml|human_readable|custom}\n"
	    << "  {-r|--repair}\n"
	    << "  {-m|--metadata-snap} [block#]\n"
	    << "  {-o <xml file>}\n"
	    << "  {--dev-id} <dev-id>\n"
	    << "  {--skip-mappings}\n"
	    << "  {-V|--version}" << endl;
}

int
thin_dump_cmd::run(int argc, char **argv)
{
	int c;
	char const *output = NULL;
	const char shortopts[] = "hm::o:f:rV";
	char *end_ptr;
	block_address metadata_snap = 0;
	uint64_t dev_id;
	struct flags flags;

	const struct option longopts[] = {
		{ "help", no_argument, NULL, 'h'},
		{ "metadata-snap", optional_argument, NULL, 'm' },
		{ "output", required_argument, NULL, 'o'},
		{ "format", required_argument, NULL, 'f' },
		{ "repair", no_argument, NULL, 'r'},
		{ "dev-id", required_argument, NULL, 1 },
		{ "skip-mappings", no_argument, NULL, 2 },
		{ "version", no_argument, NULL, 'V'},
		{ NULL, no_argument, NULL, 0 }
	};

	while ((c = getopt_long(argc, argv, shortopts, longopts, NULL)) != -1) {
		switch(c) {
		case 'h':
			usage(cout);
			return 0;

		case 'f':
			flags.format = optarg;
			break;

		case 'r':
			flags.opts.repair_ = true;
			break;

		case 'm':
			flags.use_metadata_snap = true;
			if (optarg) {
				// FIXME: deprecate this option
				metadata_snap = strtoull(optarg, &end_ptr, 10);
				if (end_ptr == optarg) {
					cerr << "couldn't parse <metadata-snap>" << endl;
					usage(cerr);
					return 1;
				}

				flags.snap_location = metadata_snap;
			}
			break;

		case 'o':
			output = optarg;
			break;

		case 1:
			dev_id = strtoull(optarg, &end_ptr, 10);
			if (end_ptr == optarg) {
				cerr << "couldn't parse <dev-id>\n";
				usage(cerr);
				return 1;
			}
			flags.opts.select_dev(dev_id);
			break;

		case 2:
			flags.opts.skip_mappings_ = true;
			break;

		case 'V':
			cout << THIN_PROVISIONING_TOOLS_VERSION << endl;
			return 0;

		default:
			usage(cerr);
			return 1;
		}
	}

	if (argc == optind) {
		cerr << "No input file provided." << endl;
		usage(cerr);
		return 1;
	}

	return dump(argv[optind], output, flags);
}

//----------------------------------------------------------------
